File: nextauth-gateway/lib/oauthContract.ts
import fs from "fs";
import path from "path";

type OAuthContract = {
  oauth_request_id: {
    regex: string;
    min_length: number;
    max_length: number;
    error_codes: {
      invalid: string;
      unknown: string;
    };
  };
  metrics: {
    invalid_state_rejected: string;
    unknown_oauth_request_id: string;
  };
};

let cached: OAuthContract | null = null;

function loadContract(): OAuthContract {
  if (cached) {
    return cached;
  }
  const contractPath = path.resolve(process.cwd(), "..", "contract", "oauth_contract.json");
  const raw = fs.readFileSync(contractPath, "utf-8");
  cached = JSON.parse(raw) as OAuthContract;
  return cached;
}

export function getOAuthRequestIdRegex(): RegExp {
  const { regex } = loadContract().oauth_request_id;
  return new RegExp(regex);
}

export function getOAuthErrorCode(code: "invalid" | "unknown"): string {
  return loadContract().oauth_request_id.error_codes[code];
}

export function getOAuthMetric(metric: "invalid_state_rejected" | "unknown_oauth_request_id"): string {
  return loadContract().metrics[metric];
}

File: app/services/oauth_contract.py
import json
from pathlib import Path
from typing import Any


_CONTRACT_PATH = Path(__file__).resolve().parents[2] / "contract" / "oauth_contract.json"
_CONTRACT_CACHE: dict[str, Any] | None = None


def _load_contract() -> dict[str, Any]:
    global _CONTRACT_CACHE
    if _CONTRACT_CACHE is None:
        with _CONTRACT_PATH.open("r", encoding="utf-8") as handle:
            _CONTRACT_CACHE = json.load(handle)
    return _CONTRACT_CACHE


def get_oauth_request_id_regex() -> str:
    return _load_contract()["oauth_request_id"]["regex"]


def get_oauth_error_code(code: str) -> str:
    return _load_contract()["oauth_request_id"]["error_codes"][code]


def get_oauth_metric(metric: str) -> str:
    return _load_contract()["metrics"][metric]
